crates/store/src/server/api.rs

```diff
@@ -275,32 +274,28 @@ impl api_server::Api for StoreApi {
                     .try_into()
                     .map_err(|err: ConversionError| Status::invalid_argument(err.to_string()))?;
 
-                match (account_state.account_id.is_on_chain(), account_update.details.is_some()) {
-                    (true, false) => {
+                let details = AccountUpdateDetails::read_from_bytes(&account_update.details)
+                    .map_err(|err| Status::invalid_argument(err.to_string()))?;
+
+                match (account_state.account_id.is_on_chain(), details.is_private()) {
+                    (true, true) => {
                         return Err(Status::invalid_argument("On-chain account must have details"));
                     },
-                    (false, true) => {
+                    (false, false) => {
```

Comment:
I think in a follow up we should introduce enums to make this easier to read, something like:

```suggestion
                    (State::OnChain, Account::Private) => {
```

(I haven't put much thought on the naming, just sharing the desire to make the code better)